### PR TITLE
more resiliancy

### PR DESCRIPTION
**Pull request by mr-c**
Experimental fix for the following errors on the EMBL Slurm cluster:

```
  File "/g/korbel2/weber/toil_venv/lib/python3.10/site-packages/toil/fileStores/nonCachingFileStore.py", line 88, in open
    os.remove(self.jobStateFile)
FileNotFoundError: [Errno 2] No such file or directory: '/run/user/25221/toil/dc0f5af078645052917223b61a7fd1fd/tmp6fvfq2kz.jobState'
```

```
Exception in thread Thread-2 (statsAndLoggingAggregator):
Traceback (most recent call last):
  File "/g/easybuild/x86_64/Rocky/8/haswell/software/Python/3.10.4-GCCcore-11.3.0/lib/python3.10/threading.py", line 1009, in _bootstrap_inner
    self.run()
  File "/g/easybuild/x86_64/Rocky/8/haswell/software/Python/3.10.4-GCCcore-11.3.0/lib/python3.10/threading.py", line 946, in run
    self._target(*self._args, **self._kwargs)
  File "/g/korbel2/weber/toil_venv/lib/python3.10/site-packages/toil/statsAndLogging.py", line 180, in statsAndLoggingAggregator
    if jobStore.read_logs(callback) == 0:
  File "/g/korbel2/weber/toil_venv/lib/python3.10/site-packages/toil/jobStores/fileJobStore.py", line 716, in read_logs
    callback(fH)
  File "/g/korbel2/weber/toil_venv/lib/python3.10/site-packages/toil/statsAndLogging.py", line 152, in callback
    stats = json.loads(statsStr, object_hook=Expando)
  File "/g/easybuild/x86_64/Rocky/8/haswell/software/Python/3.10.4-GCCcore-11.3.0/lib/python3.10/json/__init__.py", line 359, in loads
    return cls(**kw).decode(s)
  File "/g/easybuild/x86_64/Rocky/8/haswell/software/Python/3.10.4-GCCcore-11.3.0/lib/python3.10/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/g/easybuild/x86_64/Rocky/8/haswell/software/Python/3.10.4-GCCcore-11.3.0/lib/python3.10/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

## Changelog Entry
To be copied to the [draft changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog) by merger:

 * PR submitter writes their recommendation for a changelog entry here

## Reviewer Checklist

 * [ ] Make sure it is coming from `issues/XXXX-fix-the-thing` in the Toil repo, or from an external repo.
    * [ ] If it is coming from an external repo, make sure to pull it in for CI with:
        ```
        contrib/admin/test-pr otheruser theirbranchname issues/XXXX-fix-the-thing
        ```
    * [ ] If there is no associated issue, [create one](https://github.com/DataBiosphere/toil/issues/new).
* [ ] Read through the code changes. Make sure that it doesn't have:
    * [ ] Addition of trailing whitespace.
    * [ ] New variable or member names in `camelCase` that want to be in `snake_case`.
    * [ ] New functions without [type hints](https://docs.python.org/3/library/typing.html).
    * [ ] New functions or classes without informative docstrings.
    * [ ] Changes to semantics not reflected in the relevant docstrings.
    * [ ] New or changed command line options for Toil workflows that are not reflected in `docs/running/{cliOptions,cwl,wdl}.rst` 
    * [ ] New features without tests.
* [ ] Comment on the lines of code where problems exist with a review comment. You can shift-click the line numbers in the diff to select multiple lines.
* [ ] Finish the review with an overall description of your opinion.

## Merger Checklist

* [ ] Make sure the PR passes tests.
* [ ] Make sure the PR has been reviewed **since its last modification**. If not, review it.
* [ ] Merge with the Github "Squash and merge" feature.
    * [ ] If there are multiple authors' commits, add [Co-authored-by](https://github.blog/2018-01-29-commit-together-with-co-authors/) to give credit to all contributing authors.
* [ ] Copy its recommended changelog entry to the [Draft Changelog](https://github.com/DataBiosphere/toil/wiki/Draft-Changelog).
* [ ] Append the issue number in parentheses to the changelog entry.

### CODE REVIEW

**Comment by adamnovak** [comment]
I feel like this might end up masking a real problem with the job state files; we aren't *supposed* to be able to lose them if the filesystem is a properly consistent one.

Do you happen to have more logs from the problems EMBL is reporting?

Maybe we should be logging warnings/errors when these files are missing so that we can keep searching for the underlying problem?

File: src/toil/fileStores/nonCachingFileStore.py [hunk 1]
```diff
@@ -85,7 +85,10 @@ def open(self, job: Job) -> Generator[None, None, None]:
                 self.logToMaster(disk_usage, level=logging.DEBUG)
             os.chdir(startingDir)
             # Finally delete the job from the worker
-            os.remove(self.jobStateFile)
+            try:
+                os.remove(self.jobStateFile)
+            except FileNotFoundError:
+                pass
```
Comment (adamnovak) [inline src/toil/fileStores/nonCachingFileStore.py:91]:
```suggestion
                logger.exception('Job state file %s has gone missing unexpectedly; some cleanup for failed jobs may be getting skipped!', self.jobStateFile)
                pass
```

File: src/toil/statsAndLogging.py [hunk 2]
```diff
@@ -148,6 +148,8 @@ def callback(fileHandle: Union[IO[bytes], IO[str]]) -> None:
             if not isinstance(statsStr, str):
                 statsStr = statsStr.decode()
             stats = json.loads(statsStr, object_hook=Expando)
+            if not stats:
```
Comment (adamnovak) [inline src/toil/statsAndLogging.py:151]: I'm not sure that an empty stats is disallowed, so just skipping them silently is *probably* fine?

**Review by adamnovak — APPROVED** [review]
I would add this message here but otherwise this seems fine.

**Comment by mr-c** [comment]
@adamnovak the other option is to add a limited retry loop. Several of the filesystems here are very inconsistent.

**Comment by mr-c** [comment]
@adamnovak , yeah, I was really surprised. At least on the leader node (which was a login/submission node for the cluster), `/var/run` was a `tmpfs` filesystem.

Maybe @weber8thomas can confirm, or help us with troubleshooting? (I'm going back to Berlin today)

**Comment by adamnovak** [comment]
@mr-c Is this ready to de-draftify or are you still working on it?

**Comment by adamnovak** [comment]
I've managed to hit what I think is the same problem on UCSC's new Slurm cluster; the non-caching file store's job state files are being destroyed before they are supposed to be. I'm now trying to work out if it is really Toil doing this, or if Slurm is doing something weird with /var/run that causes this.

**Comment by adamnovak** [comment]
OK, on our Slurm cluster, it seems that jobs get an `XDG_RUNTIME_DIR` that is just copied from the environment on the head node, and which *doesn't* mean that that directory is actually available as an *actual* XDG_RUNTIME_DIR to the current Slurm job? For some reason the directory will often exist, but will get cleaned up after a bit by the system even if Toil has been trying to use it to store stuff.

**Comment by mr-c** [comment]
> OK, on our Slurm cluster, it seems that jobs get an `XDG_RUNTIME_DIR` that is just copied from the environment on the head node, and which *doesn't* mean that that directory is actually available as an *actual* XDG_RUNTIME_DIR to the current Slurm job? For some reason the directory will often exist, but will get cleaned up after a bit by the system even if Toil has been trying to use it to store stuff.

That sounds worth reporting to the Slurm people

**Pull request merged**